### PR TITLE
Improvement: Improvement prime tower

### PR DESCRIPTION
**Pull request by wen-qing-255**

### CODE REVIEW

File: src/app/flux/printing/index.js [hunk 1]
```diff
@@ -1121,11 +1124,21 @@ export const actions = {
...
         const currentModelName = path.basename(models[0]?.modelName, path.extname(models[0]?.modelName));
         const renderGcodeFileName = `${currentModelName}_${new Date().getTime()}`;
+        const hasPrimeTower = (printingToolhead === DUAL_EXTRUDER_TOOLHEAD_FOR_SM2 && activeDefinition.settings.prime_tower_enable.default_value);
         // Prepare definition file
-        const { size } = getState().machine;
         await dispatch(actions.updateActiveDefinitionMachineSize(size));
-
-        const finalDefinition = definitionManager.finalizeActiveDefinition(activeDefinition);
+        if (hasPrimeTower) {
+            const modelGroupBBox = modelGroup._bbox;
```
Comment (jane-rose) [inline src/app/flux/printing/index.js:1131]: just use `new Three.Box3 `

Comment (wen-qing-255) [reply src/app/flux/printing/index.js:1131]: never modify modelGroupBBox，need platform boundingbox to get prime tower position.

File: src/app/flux/printing/index.js [hunk 2]
```diff
@@ -780,6 +780,7 @@ export const actions = {
             activeDefinition.settings.support_roof_line_width.default_value = extruderDef.settings.support_roof_line_width.default_value;
             activeDefinition.settings.support_bottom_line_width.default_value = extruderDef.settings.support_bottom_line_width.default_value;
             activeDefinition.settings.prime_tower_line_width.default_value = extruderDef.settings.prime_tower_line_width.default_value;
+            activeDefinition.settings.prime_tower_wipe_enabled.default_value = true;
```
Comment (bbaa77770) [inline src/app/flux/printing/index.js:783]: Why is always true

File: src/app/flux/printing/index.js [hunk 3]
```diff
@@ -1121,11 +1124,21 @@ export const actions = {
...
         const renderGcodeFileName = `${currentModelName}_${new Date().getTime()}`;
+        const hasPrimeTower = (printingToolhead === DUAL_EXTRUDER_TOOLHEAD_FOR_SM2 && activeDefinition.settings.prime_tower_enable.default_value);
         // Prepare definition file
-        const { size } = getState().machine;
         await dispatch(actions.updateActiveDefinitionMachineSize(size));
-
-        const finalDefinition = definitionManager.finalizeActiveDefinition(activeDefinition);
+        if (hasPrimeTower) {
+            const modelGroupBBox = modelGroup._bbox;
+            const primeTowerModel = lodashFind(modelGroup.getModels(), { type: 'primeTower' });
```
Comment (xiekun1992) [inline src/app/flux/printing/index.js:1132]: consider to define a method in `modelGroup` to filter out specific sub models

File: src/app/flux/printing/index.js [hunk 4]
```diff
@@ -1173,7 +1186,7 @@ export const actions = {
             // Use setTimeout to force export executes in next tick, preventing block of updateState()
 
             setTimeout(async () => {
-                const models = modelGroup.models.filter(i => i.visible).reduce((pre, model) => {
+                const models = modelGroup.models.filter(i => i.visible && i.type !== 'primeTower').reduce((pre, model) => {
```
Comment (xiekun1992) [inline src/app/flux/printing/index.js:1189]: consider to define a method in `modelGroup` to filter out specific sub models

File: src/app/flux/printing/index.js [hunk 5]
```diff
@@ -1098,8 +1100,9 @@ export const actions = {
             definitionId: 'snapmaker_extruder_1'
         });
 
-        const models = filter(modelGroup.getModels(), { 'visible': true });
-
+        const models = filter(modelGroup.getModels(), (modelItem) => {
```
Comment (xiekun1992) [inline src/app/flux/printing/index.js:1103]: consider to define a method in `modelGroup` to filter out specific sub models

**Review by xiekun1992 — CHANGES_REQUESTED** [review]

File: src/app/flux/printing/index.js [hunk 6]
```diff
@@ -1098,8 +1100,9 @@ export const actions = {
...
+            return modelItem.visible && modelItem.type !== 'primeTower';
+        });
```
Comment (jane-rose) [inline src/app/flux/printing/index.js:1105]: add function inside modelGroup to filter models

**Review by jane-rose — APPROVED** [review]

**Pull request merged**